independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarResource.java

```diff
@@ -106,20 +124,8 @@ public ManifestInfo getManifestInfo() {
     }
 
     @Override
-    public ProtectionDomain getProtectionDomain(ClassLoader classLoader) {
-        final URL url;
-        try {
-            String path = jarPath.toAbsolutePath().toString();
-            if (!path.startsWith("/")) {
-                path = '/' + path;
-            }
-            URI uri = new URI("file", null, path, null);
-            url = uri.toURL();
-        } catch (URISyntaxException | MalformedURLException e) {
-            throw new RuntimeException("Unable to create protection domain for " + jarPath, e);
-        }
-        CodeSource codesource = new CodeSource(url, (Certificate[]) null);
-        return new ProtectionDomain(codesource, null, classLoader, null);
+    public ProtectionDomain getProtectionDomain() {
+        return protectionDomain;
```

Comment:
this might need some thoughts regading threadsafety?

Maybe you could simplify things by maintaining the same API as it had, then lazily initialize the ProtectionDomain domain on first request. This way you can also check if the classloader is the same as the one you have cached the `ProtectionDomain` for, and in case rebuild the instance.

This would ensure there's no way to get the wrong `ProtectionDomain`, for example in case the same `JarResource` was ever used by multiple / different classloader instances - I'm thinking for example it might be reusing the same `JarResource` by a fresh classloader on hot reload?

Reply:
Good point, but this ClassLoader is only used for `fast-jar`, it's **not** the ClassLoader that is used in dev-mode, so  no hot deployment can happen.

There shouldn't be any thread safety issues whatsoever, because the `init` method which creates the `protection` domain is only called once from the `main` thread.

Otherwise I am missing something from what you said :)

Reply:
the `init` method is called once, but `getProtectionDomain` is invoked by any thread, as we support parrallel classloading. At very least the field should be volatile ?

Reply:
and once you made it volatile, you might as well make it lazy so to not need to remember to invoke init

Reply:
`getProtectionDomain` can indeed be called by multiple threads, but since we are just returning a value that has been pre-computed (in `init`), do we really need to make it volatile?

Reply:
yes - another thread will not necessarily be able to "see" its state otherwise, as there's no barrier at all giving you such guarantee.

Reply:
OK, I'll make it volatile then

Reply:
> and once you made it volatile, you might as well make it lazy so to not need to remember to invoke init

I don't think that is the case - because `init` passes in the `ClassLoader` which is needed to create the `ProtectionDomain`. 
If we were to remove `init` and go back to the assumption that `ClassLoader` resource could be used for multiple ClassLoaders, then we would effectively need to reintroduce the map.

Am I missing something here?

Reply:
We could of course make `protectionDomain` lazy and keep hold of the `ClassLoader`, but then that would mean adding the ClassLoader field as volatile as well.

Reply:
Ok, might be fine to use the `init` method and initialize this eagerly then.